### PR TITLE
Change: use tree-kill

### PR DESCRIPTION
**Pull request by sevenc-nanashi**
This PR makes this plugin use `tree-kill`, which kills child processes.

### CODE REVIEW

**Comment by caoxiemeihao** [comment]
Is this required? Do the subprocesses refer to the ones opened in Electron Main? Or is there another way to accomplish what you need?
`vite-plugin-electron` needs to be kept lean enough.

**Comment by sevenc-nanashi** [comment]
Yes, tree-kill kills the subprocesses of electron main.
We can kill the process with SIGINT and do post-process in app, but app may refuse SIGINT and may do something bad in vite-plugin-electron.

File: package.json [hunk 1]
```diff
@@ -21,11 +21,14 @@
     "build": "tsc --emitDeclarationOnly && vite build",
     "prepublishOnly": "npm run build"
   },
-  "dependencies": {},
+  "dependencies": {
+    "tree-kill": "^1.2.2",
+    "vite-plugin-electron": "link:"
```
Comment (caoxiemeihao) [inline package.json:26]: What is the reason for the change here?

**Comment by caoxiemeihao** [comment]
Does this happen often? Not sure there is a good benefit from introducing tree kill, or is there no way to solve this problem without introducing tree kill?

File: package.json (hunk 1 above)
Comment (sevenc-nanashi) [reply package.json:26]: idk, I'm reverting it

**Comment by sevenc-nanashi** [comment]
There's a issue to do tree-kill in node, but it's closed: https://github.com/nodejs/node/issues/40438

**Comment by caoxiemeihao** [comment]
Once again, thank you for the PR.
But vite-plugin-electron goes for minimalism and zero dependencies is hard to come by.
Not sure how many users will need this feature, and even as it is now, users can still do the same thing in `onstart`.
I'll keep an eye on this PR.
https://github.com/electron-vite/vite-plugin-electron/blob/v0.11.1/examples/custom-start-electron-app/vite.config.ts#L8-L11

**Review by Sammiegirl411 — APPROVED** [review]

**Review by Sammiegirl411 — APPROVED** [review]

**Review by Sammiegirl411 — APPROVED** [review]

**Review by Sammiegirl411 — APPROVED** [review]

**Review by Sammiegirl411 — APPROVED** [review]

**Comment by caoxiemeihao** [comment]
Okay!
Looks like other people are having this issue too, so thanks for the PR.

But I don't like to add [tree-kill](https://www.npmjs.com/package/tree-kill) to the required dependencies, like [vite-plugin-electron-renderer](https://github.com/electron-vite/vite-plugin-electron-renderer)

**package.json**

```diff
{
  "peerDependencies": {
    "vite-plugin-electron-renderer": "*",
+   "tree-kill": "*"
  },
  "peerDependenciesMeta": {
    "vite-plugin-electron-renderer": {
      "optional": true
    },
+   "tree-kill": {
+     "optional": true
+   }
  }
}
```

**src/index.ts**

```ts
startup.exit = async () => {
  if (process.electronApp) {
    process.electronApp.removeAllListeners()
    try {
      const { default: treeKill } = await import('tree-kill')
      treeKill(process.electronApp.pid!)
    } catch (error) {
      process.electronApp.kill()
    }
  }
}
```

Would you mind making some changes? 😄

**cc** @shuyuTech, thanks you for explaining this issue. 👍

**Comment by sevenc-nanashi** [comment]
Done.

**Comment by sevenc-nanashi** [comment]
wtf

**Comment by sevenc-nanashi** [comment]
Managed to open again.

**Comment by huangxulei** [comment]
你好，我是用js的，dev有时候,electron启动不起来。 然后随便修改下无关代码，重新dev，又能起来。不知道什么原因。

**Comment by caoxiemeihao** [comment]
> 你好，我是用js的，dev有时候,electron启动不起来。 然后随便修改下无关代码，重新dev，又能起来。不知道什么原因。

把 tree-kill 去掉试试先，这副作用真的 yue 🤮

**Comment by huangxulei** [comment]
没注意，原来是vite-plugin-electron里面添加了tree-kill，是可以在哪来设置可以关掉吗？

**Comment by imblowsnow** [comment]
> > 你好，我是用js的，dev有时候,electron启动不起来。 然后随便修改下无关代码，重新dev，又能起来。不知道什么原因。
> 
> 把 tree-kill 去掉试试先，这副作用真的 yue 🤮

我也碰到这个问题了，突然就起不来了

**Comment by huangxulei** [comment]
> > > 你好，我是用js的，dev有时候,electron启动不起来。 然后随便修改下无关代码，重新dev，又能起来。不知道什么原因。
> > 
> > 
> > 把 tree-kill 去掉试试先，这副作用真的 yue 🤮
> 
> 我也碰到这个问题了，突然就起不来了

我调试了代码 dev 的时候 , 会自动打包, 然后会启动一个closeBundle , 然后我console了, 你可以看我后面贴的代码.
空的时候研究一下，我现在用用能启动就不关了。就没理他。不过好像最新代码里面已经添加了 tree-kill, 不知道怎么关闭这个功能. 

 closeBundle() {
                  if (++closeBundleCount < entryCount)
                    return;
                  if (options2.onstart) {// **走这里就启动不了.**
                    options2.onstart.call(this, {
                      startup,
                      reload() {
                        server.ws.send({ type: "full-reload" });
                      }
                    });
                  } else {    //**这里是启动electron**
                    startup();
                  }
                }

**Comment by caoxiemeihao** [comment]
哈哈，感谢你的提醒。
晚点我调试下试试，源码没多少，大家都能搞

**Comment by noob9527** [comment]
> huangxulei

我也遇到一样的问题
```
  window.once('ready-to-show', () => {
    logger.debug('ready-to-show')
    window.show()
  })
```
我是监听 ready-to-show 事件启动 window 的，但在 dev 环境下，经常不触发这个事件。
随便改改代码，加一个空行之类的又可以了。
但我没有装 `tree-kill`
